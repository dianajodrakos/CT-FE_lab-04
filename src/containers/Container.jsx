import React, { Component } from 'react';
import Controls from '../components/api-client/Controls';
import Display from '../components/api-client/Display';
import Header from '../components/Header';
import History from '../components/history/History';
import { callAPI } from '../services/apiServices';
import styles from '../index.css';

export default class Container extends Component {
state = {
  loading: false,
  url: '',
  method: '',
  body: '',
  history: [],
  display: { 'Hello, Dave': 'You\'re looking well today.' }
}

handleChange = (event) => {
  this.setState({ [event.target.name]: event.target.value });
}

handleSubmit = async (event) => {
  event.preventDefault();
  this.setState({ loading: true });

  const display = await callAPI(
    this.state.url, 
    this.state.method, 
    this.state.body
  );

  this.setState({ loading: false, display });

}

// handleClick = (event) => {
  
// }

render() {
  console.log(this.state);

  const { loading, display } = this.state;
  return (
    <>
      <Header />
      <section className={styles.main}>
        <History />
        <div>
          <Controls onChange={this.handleChange} onSubmit={this.handleSubmit}/>
          { loading 
            ? <div>Loading...</div>
            :<Display display={display}/> 
          }
        </div>
      </section>
    </>
  );
}
}
