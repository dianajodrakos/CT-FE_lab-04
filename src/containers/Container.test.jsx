/* eslint-disable max-len */
import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { rest } from 'msw';
import { setupServer } from 'msw/node';
import listData from '../fixtures/listData.json';
import Container from './Container';

const server = setupServer(
  rest.get('https://ac-vill.herokuapp.com/villagers/', (req, res, ctx) => {
    return res(ctx.json(listData));
  })
);

describe('Container test', () => {
  beforeAll(() => server.listen());
  afterAll(() => server.close());

  it('should enter a URL and request method, click submit, and return a response from an API', async () => {
    const container = render(<Container />);

    const controls = screen.getByRole('form', { name: 'REST controls' });
    const display = screen.getByRole('article', { name: 'response' });

    expect(controls).toMatchSnapshot();

    screen.getByText('REST Clone');

    const input = await screen.findByPlaceholderText('URL');
    userEvent.type(input, 'https://ac-vill.herokuapp.com/villagers/');

    const radioButton = await screen.findByLabelText('GET');
    userEvent.click(radioButton);

    const submitButton = await screen.findByRole('button', { name: 'submit' });

    userEvent.click(submitButton);

    return waitFor(() => {
      const response = screen.getByText('"Admiral"', { exact: false });

      expect(response).toBeInTheDocument();
      expect(container).toMatchSnapshot();
      expect(display).toMatchSnapshot();
    });
  });
});
