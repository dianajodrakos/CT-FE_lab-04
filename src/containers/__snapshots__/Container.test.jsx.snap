// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Container test should return a call to an API 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="rooot"
      >
        <h1>
          REST Clone
        </h1>
        <section
          class="main"
        >
          <div
            class="history"
          >
            <h2>
              Call History
            </h2>
          </div>
          <div>
            <form
              class="display"
            >
              <input
                aria-label="url"
                name="url"
                placeholder="URL"
                type="text"
              />
              <section>
                <div>
                  <label
                    for="get"
                  >
                    <input
                      id="get"
                      name="method"
                      type="radio"
                      value="GET"
                    />
                    <p>
                      GET
                    </p>
                  </label>
                  <label
                    for="post"
                  >
                    <input
                      id="post"
                      name="method"
                      type="radio"
                      value="POST"
                    />
                    <p>
                      POST
                    </p>
                  </label>
                  <label
                    for="put"
                  >
                    <input
                      id="put"
                      name="method"
                      type="radio"
                      value="PUT"
                    />
                    <p>
                      PUT
                    </p>
                  </label>
                  <label
                    for="patch"
                  >
                    <input
                      id="patch"
                      name="method"
                      type="radio"
                      value="PATCH"
                    />
                    <p>
                      PATCH
                    </p>
                  </label>
                  <label
                    for="delete"
                  >
                    <input
                      id="delete"
                      name="method"
                      type="radio"
                      value="DELETE"
                    />
                    <p>
                      DELETE
                    </p>
                  </label>
                </div>
                <button
                  aria-label="submit"
                >
                  Submit
                </button>
              </section>
              <textarea
                aria-label="body"
                name="body"
                placeholder="raw JSON body"
              />
            </form>
            <div
              class="display"
            >
              Loading...
            </div>
          </div>
        </section>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="rooot"
    >
      <h1>
        REST Clone
      </h1>
      <section
        class="main"
      >
        <div
          class="history"
        >
          <h2>
            Call History
          </h2>
        </div>
        <div>
          <form
            class="display"
          >
            <input
              aria-label="url"
              name="url"
              placeholder="URL"
              type="text"
            />
            <section>
              <div>
                <label
                  for="get"
                >
                  <input
                    id="get"
                    name="method"
                    type="radio"
                    value="GET"
                  />
                  <p>
                    GET
                  </p>
                </label>
                <label
                  for="post"
                >
                  <input
                    id="post"
                    name="method"
                    type="radio"
                    value="POST"
                  />
                  <p>
                    POST
                  </p>
                </label>
                <label
                  for="put"
                >
                  <input
                    id="put"
                    name="method"
                    type="radio"
                    value="PUT"
                  />
                  <p>
                    PUT
                  </p>
                </label>
                <label
                  for="patch"
                >
                  <input
                    id="patch"
                    name="method"
                    type="radio"
                    value="PATCH"
                  />
                  <p>
                    PATCH
                  </p>
                </label>
                <label
                  for="delete"
                >
                  <input
                    id="delete"
                    name="method"
                    type="radio"
                    value="DELETE"
                  />
                  <p>
                    DELETE
                  </p>
                </label>
              </div>
              <button
                aria-label="submit"
              >
                Submit
              </button>
            </section>
            <textarea
              aria-label="body"
              name="body"
              placeholder="raw JSON body"
            />
          </form>
          <div
            class="display"
          >
            Loading...
          </div>
        </div>
      </section>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
